data-prepper-core/src/integrationTest/java/com/amazon/dataprepper/integration/EndToEndServiceMapTest.java

```diff
@@ -172,44 +194,51 @@ public static ExportTraceServiceRequest getExportTraceServiceRequest(List<Resour
                     spanKind
             );
             spansList.add(rs);
-            spanIdToRS.put(spanId, rs);
+        }
+        return spansList;
+    }
+
+    private List<Map<String, Object>> getPossibleEdges(final String traceId, final List<ServiceMapTestData> data) {
+        final Map<String, ServiceMapTestData> spanIdToServiceMapTestData = data.stream()
+                .collect(Collectors.toMap(serviceMapTestData -> serviceMapTestData.spanId, serviceMapTestData -> serviceMapTestData));
+        final List<Map<String, Object>> possibleEdges = new ArrayList<>();
+        for (final ServiceMapTestData currData : data) {
+            final String parentId = currData.parentId;
             if (parentId != null) {
-                ResourceSpans parentRS = spanIdToRS.get(parentId);
-                String parentServiceName = parentRS.getResource().getAttributes(0).getValue().getStringValue();
-                if (!parentServiceName.equals(serviceName)) {
-                    String rootSpanName = getRootSpanName(parentId);
+                final ServiceMapTestData parentData = spanIdToServiceMapTestData.get(parentId);
+                if (parentData != null && !parentData.serviceName.equals(currData.serviceName)) {
+                    String rootSpanName = getRootSpanName(parentId, spanIdToServiceMapTestData);
                     Map<String, Object> destination = new HashMap<>();
-                    destination.put("resource", spanName);
-                    destination.put("domain", serviceName);
+                    destination.put("resource", currData.name);
+                    destination.put("domain", currData.serviceName);
                     Map<String, Object> edge = new HashMap<>();
-                    edge.put("serviceName", parentServiceName);
-                    edge.put("kind", parentRS.getInstrumentationLibrarySpans(0).getSpans(0).getKind().name());
+                    edge.put("serviceName", parentData.serviceName);
+                    edge.put("kind", parentData.spanKind.name());
                     edge.put("traceGroupName", rootSpanName);
                     edge.put("destination", destination);
                     edge.put("target", null);
                     possibleEdges.add(edge);
```

Comment:
Nit: Nothing incorrect here, but I think this code could probably use some cleanup in terms of separating large chunks of code into logical functions. For example:

getEdgeMap(curData, parentData)

Reply:
Will take another look.